64_EmulatedFloatTest/main.cpp

```diff
@@ -1033,6 +1047,292 @@ class CompatibilityTest final : public MonoDeviceApplication, public MonoAssetMa
+            const IQueue::SSubmitInfo::SCommandBufferInfo cmdbufsBegin[] = { {.cmdbuf = m_timestampBeforeCmdBuff.get()} };
+            beforeTimestapSubmitInfo[0].commandBuffers = cmdbufsBegin;
+            IQueue::SSubmitInfo::SSemaphoreInfo beforeTimestampSignals[] = { {.semaphore = m_beforeTimestampSemaphore.get(), .value = ++m_beforeTimestampSemaphoreCounter, .stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT} };
+            beforeTimestapSubmitInfo[0].signalSemaphores = beforeTimestampSignals;
+
+            IQueue::SSubmitInfo afterTimestapSubmitInfo[1] = {};
+            const IQueue::SSubmitInfo::SCommandBufferInfo cmdbufsEnd[] = { {.cmdbuf = m_timestampAfterCmdBuff.get()} };
+            afterTimestapSubmitInfo[0].commandBuffers = cmdbufsEnd;
+            IQueue::SSubmitInfo::SSemaphoreInfo afterTimestapSignals[] = { {.semaphore = m_afterTimestampSemaphore.get(), .value = ++m_afterTimestampSemaphoreCounter, .stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT} };
+            afterTimestapSubmitInfo[0].signalSemaphores = afterTimestapSignals;
+
+            IQueue::SSubmitInfo benchmarkSubmitInfos[1] = {};
+            const IQueue::SSubmitInfo::SCommandBufferInfo cmdbufs[] = { {.cmdbuf = m_cmdbuf.get()} };
+            benchmarkSubmitInfos[0].commandBuffers = cmdbufs;
+            IQueue::SSubmitInfo::SSemaphoreInfo signals[] = { {.semaphore = m_semaphore.get(), .value = ++m_semaphoreCounter, .stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT} };
+            benchmarkSubmitInfos[0].signalSemaphores = signals;
+
+            m_pushConstants.benchmarkMode = mode;
+            recordCmdBuff();
+
+            // warmup runs
+            for (int i = 0; i < WarmupIterations; ++i)
+            {
+                m_computeQueue->submit(benchmarkSubmitInfos);
+                signals[0].value = ++m_semaphoreCounter;
+            }
+
+            m_device->waitIdle();
+            m_computeQueue->submit(beforeTimestapSubmitInfo);
+            m_device->waitIdle();
+
+            // actual benchmark runs
+            for (int i = 0; i < Iterations; ++i)
+            {
+                m_computeQueue->submit(benchmarkSubmitInfos);
+                signals[0].value = ++m_semaphoreCounter;
+            }
+
+            m_device->waitIdle();
+            m_computeQueue->submit(afterTimestapSubmitInfo);
+            m_device->waitIdle();
+
+            beforeTimestampSignals[0].value = ++m_beforeTimestampSemaphoreCounter;
+            afterTimestapSignals[0].value = ++m_afterTimestampSemaphoreCounter;
+
+            const uint64_t nativeBenchmarkTimeElapsedNanoseconds = calcTimeElapsed();
+            const float nativeBenchmarkTimeElapsedSeconds = double(nativeBenchmarkTimeElapsedNanoseconds) / 1000000000.0;
+
+            m_logger->log("%llu ns, %f s", ILogger::ELL_PERFORMANCE, nativeBenchmarkTimeElapsedNanoseconds, nativeBenchmarkTimeElapsedSeconds);
+        }
```

Comment:
If I understood  correctly you're having 3 command buffers (beforeTimeStamp, MainWork, afterTimestamp) and you're submitting them and ensuring the order by waitIdle instead of semaphores.
and You're submitting the MainWork Iteration Times (1000), It makes sense to do that instead of recording a command buffer with 1000 dispatches inside, I get that you also want to keep reusing your command buffer.

waitIdle syncs with CPU and then submits back to GPU, to get better timings you need to extract your first and last  mainWork submits and connect then to the timeStamp submits via semaphores instead of waitIdle, so first submit wait for beforeTimestamp, and afterTimestamp waits for last submit. this should be easy to achieve by doing you loop Iterations-1 times, and writing 2 lines of waitSemaphore codes 🙂